### ISSUE TITLE
Performance issue if `plus` is expensive

### ISSUE THREAD

**Issue by nevillelyh** [spotify/featran#23]
For transformers like `*HotEncoder` where `plus` expensive, i.e. set or map concatenation, we might end up creating many temporary objects and causing GC pressure. We could potentially override `sumOption` but not sure how to implement this across multiple transformers.

**Comment by richwhitjr** [comment]
Maybe we should have a one-hot-encoder which is just a hasher.  Given a possible hash space it takes the category and hashes into the space.  It would basically be free hashing if a few collisions are OK.  Another option is that we can count distinct elements using a HLL and use that to create the hash space.

**Comment by nevillelyh** [comment]
@richwhitjr can you post some references here? I guess collision is unavoidable and HLL needs to be a separate step outside of featran.

**Comment by richwhitjr** [comment]
I meant to implement a transformer that matches this https://www.tensorflow.org/api_docs/python/tf/contrib/layers/sparse_column_with_hash_bucket.  It requires a hash_bucket_size for TF so in our case we could either let the user provide that value or maybe optionally attempt to estimate it based on a HLL in the aggregation step,.

**Comment by nevillelyh** [comment]
TF's implementation here:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/string_to_hash_bucket_op.h
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/string_to_hash_bucket_op.cc
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/lib/hash/hash.h

**Comment by nevillelyh** [comment]
I'll try port those over.

**Comment by nevillelyh** [comment]
Have something WIP here: https://github.com/spotify/featran/tree/neville/hash-hot

The collision rate & output dimension can be tuned via `hashBucketSize` but could be pretty high though.

Using a list of 466544 English words, HLL estimates size at 464220. Here're the collisions stats at different sizes:

```
Input size: 466544
numBucketSize: 1073741824, collisions: 97
numBucketSize: 536870912, collisions: 210
numBucketSize: 268435456, collisions: 398
numBucketSize: 134217728, collisions: 836
numBucketSize: 67108864, collisions: 1629
numBucketSize: 33554432, collisions: 3189
numBucketSize: 16777216, collisions: 6271
numBucketSize: 8388608, collisions: 12354
numBucketSize: 4194304, collisions: 24001
numBucketSize: 2097152, collisions: 44540
numBucketSize: 1048576, collisions: 77568
numBucketSize: 524288, collisions: 117448
```

**Comment by richwhitjr** [comment]
Hard to know what is acceptable.  I wonder if we can check this collision rate against TF to see what they found to be good enough.  I doubt we could do much better then theirs with this approach with going deep into hashing logic.

**Comment by nevillelyh** [comment]
Well most collisions are caused by the `hash % bucket size` step. Without that, i.e. when `size == Int.MaxValue`, 97 collisions is close to what I found in other benchmarks.
Maybe instead of using the estimated size, we can recommend some heuristics, like 8x (~5%) or 16x (~2.5%) estimated size?

**Comment by richwhitjr** [comment]
That makes sense. Maybe the options are collision rate or hash bucket size.  May have have to give a warning if they make the collision rate too small.

**Comment by nevillelyh** [comment]
Fixed in #26
